### PR TITLE
Fix of #2561. Disable + when there is no parent element.

### PR DESCRIPTION
**Pull request by cmbahadir**
The fix checks if there is a parent object and updates the "+" href according to it. 

- Not sure if the fixed places are correct and sufficient. It checks the existence of parent object in "New Bug", "New Test Case" and "New Test Plan" pages. 
- Not sure if it is also required to check on Admin/Management. If that is the case, please help me about the hiearchy of the objects.

### CODE REVIEW

File: tcms/bugs/static/bugs/js/mutable.js [hunk 1]
```diff
@@ -35,18 +35,27 @@ $(document).ready(function () {
 function populateVersion () {
   const productId = $('#id_product').val()
 
-  const href = $('#add_id_version')[0].href
-  $('#add_id_version')[0].href = href.slice(0, href.indexOf('&product'))
-  $('#add_id_version')[0].href += `&product=${productId}`
-  $('#id_version').find('option').remove()
-  updateVersionSelectFromProduct()
+  if ($('#id_product').val() === null) {
```
Comment (atodorov) [inline tcms/bugs/static/bugs/js/mutable.js:38]: this is the `productId` variable above. No need to read from DOM again.

File: tcms/bugs/static/bugs/js/mutable.js [hunk 2]
```diff
@@ -35,18 +35,27 @@ $(document).ready(function () {
...
+    $($('#add_id_version')[0]).addClass('disabled')
```
Comment (atodorov) [inline tcms/bugs/static/bugs/js/mutable.js:39]: the `[0]` part of the selector isn't needed.

File: tcms/bugs/static/bugs/js/mutable.js [hunk 3]
```diff
@@ -35,18 +35,27 @@ $(document).ready(function () {
...
+  } else {
+    $($('#add_id_version')[0]).removeClass('disabled')
+    const href = $('#add_id_version')[0].href
+    $('#add_id_version')[0].href = href.slice(0, href.indexOf('&product'))
+    $('#add_id_version')[0].href += `&product=${productId}`
+    $('#id_version').find('option').remove()
```
Comment (atodorov) [inline tcms/bugs/static/bugs/js/mutable.js:45]: If the + button becomes disabled then user can't click on it so you don't really have to safeguard much of this with a conditional. IMO the only thing that needs the conditional is the add/remove of the disabled class.

File: tcms/bugs/static/bugs/js/mutable.js [hunk 4]
```diff
@@ -35,18 +35,27 @@ $(document).ready(function () {
...
-  updateBuildSelectFromVersion()
+  if ($('#id_version').val() === null) {
+    $($('#add_id_build')[0]).addClass('disabled')
+  } else {
+    $($('#add_id_build')[0]).removeClass('disabled')
+    const href = $('#add_id_build')[0].href
+    $('#add_id_build')[0].href = href.slice(0, href.indexOf('&version'))
+    $('#add_id_build')[0].href += `&version=${productId}`
+    updateBuildSelectFromVersion()
+  }
```
Comment (atodorov) [inline tcms/bugs/static/bugs/js/mutable.js:60]: Same comments here.

**Review by atodorov — CHANGES_REQUESTED** [review]

File: tcms/bugs/static/bugs/js/mutable.js (hunk 3 above)
Comment (atodorov) [reply tcms/bugs/static/bugs/js/mutable.js:45]: Note: if product isn't selected the URL should end in `&product=null` which shouldn't be a problem if you can't click on the link.

File: tcms/testcases/static/testcases/js/mutable.js [hunk 5]
```diff
@@ -29,6 +29,11 @@ $(document).ready(function () {
...
 
 function populateProductCategory () {
+  if ($('#id_product').val() === null) {
+    $('#add_id_category').addClass('disabled')
+  } else {
+    $('#add_id_category').removeClass('disabled')
+  }
   const href = $('#add_id_category')[0].href
   $('#add_id_category')[0].href = href.slice(0, href.indexOf('&product'))
   $('#add_id_category')[0].href += `&product=${$('#id_product').val()}`
```
Comment (atodorov) [inline tcms/testcases/static/testcases/js/mutable.js:39]: Separate productId into a variable.

File: tcms/testplans/static/testplans/js/mutable.js [hunk 6]
```diff
@@ -29,6 +29,11 @@ $(document).ready(function () {
...
 
 function populateProductVersion () {
+  if ($('#id_product').val() === null) {
+    $('#add_id_version').addClass('disabled')
+  } else {
+    $('#add_id_version').removeClass('disabled')
+  }
   const href = $('#add_id_version')[0].href
   $('#add_id_version')[0].href = href.slice(0, href.indexOf('&product'))
   $('#add_id_version')[0].href += `&product=${$('#id_product').val()}`
```
Comment (atodorov) [inline tcms/testplans/static/testplans/js/mutable.js:39]: Separate productId into variable.

**Review by atodorov — CHANGES_REQUESTED** [review]
Minor changes requested.

File: tcms/bugs/static/bugs/js/mutable.js [hunk 7]
```diff
@@ -43,10 +51,16 @@ function populateVersion () {
...
+  if (versionId === null) {
+    $('#add_id_build').addClass('disabled')
+  } else {
+    $('#add_id_build').removeClass('disabled')
+  }
 
   const href = $('#add_id_build')[0].href
   $('#add_id_build')[0].href = href.slice(0, href.indexOf('&version'))
-  $('#add_id_build')[0].href += `&version=${productId}`
+  $('#add_id_build')[0].href += `&version=${versionId}`
```
Comment (atodorov) [inline tcms/bugs/static/bugs/js/mutable.js:64]: It's a bit strange that the URL for version popup ends with `product=null` but the URL for build popup ends with `version=None`. 

Not sure if a problem at the moment.

File: tcms/static/style/patternfly_override.css [hunk 8]
```diff
@@ -71,3 +71,10 @@ img {
 .property-value a {
     color: #c00; inherit !important;
 }
+
+a.disabled {
+    color: gray;
+    cursor: default;
```
Comment (atodorov) [inline tcms/static/style/patternfly_override.css:77]: this causes the cursor to become a circle icon, not really the default pointer.

File: tcms/static/style/patternfly_override.css [hunk 9]
```diff
@@ -71,3 +71,10 @@ img {
...
+    pointer-events: none;
+    text-decoration: none;
+}
```
Comment (atodorov) [inline tcms/static/style/patternfly_override.css:80]: While this class changes the buttons visually it doesn't prevent me from clicking on them. I can still click on the plus button and get a popup opened.

**Review by atodorov — CHANGES_REQUESTED** [review]

File: tcms/static/style/patternfly_override.css (hunk 9 above)
Comment (atodorov) [reply tcms/static/style/patternfly_override.css:80]: Try https://learn.jquery.com/using-jquery-core/faq/how-do-i-disable-enable-a-form-element/ (also try `attr` if `prop` doesn't work).

Comment (cmbahadir) [reply tcms/static/style/patternfly_override.css:80]: This is strange because they are disabled on Chrome. What kind of browser are you using?

Comment (cmbahadir) [reply tcms/static/style/patternfly_override.css:80]: > Try https://learn.jquery.com/using-jquery-core/faq/how-do-i-disable-enable-a-form-element/ (also try `attr` if `prop` doesn't work).

 `prop` does not work for links also `attr` removes href,  which requires to be constructed afterwards when there is a valid pop-up link.

Comment (atodorov) [reply tcms/static/style/patternfly_override.css:80]: > What kind of browser are you using?

Firefox.

Comment (cmbahadir) [reply tcms/static/style/patternfly_override.css:80]: I have tested on "Chrome", "Firefox" and "Opera" and `+` gets disabled on "New Bug", "New Test Case", "New Test Plan" pages when there is no parent item.

Need some help here to reproduce.

Do you also expect it to be disabled on "admin" page?

Comment (atodorov) [reply tcms/static/style/patternfly_override.css:80]: Must have been stale CSS/JS. Seems to work fine for me now.

**Pull request merged**